core/inspector.rs

```diff
@@ -236,20 +256,55 @@ impl JsRuntimeInspector {
     self.sessions.borrow().has_blocking_sessions()
   }
 
-  fn poll_sessions(
-    &self,
-    mut invoker_cx: Option<&mut Context>,
-  ) -> Result<Poll<()>, BorrowMutError> {
-    // The futures this function uses do not have re-entrant poll() functions.
-    // However it is can happpen that poll_sessions() gets re-entered, e.g.
-    // when an interrupt request is honored while the inspector future is polled
-    // by the task executor. We let the caller know by returning some error.
-    let mut sessions = self.sessions.try_borrow_mut()?;
+  fn poll_sessions_sync(&self) {
+    let (prev_poll_state, mut prev_task_waker) = self.waker.update(|w| {
+      let prev_poll_state = replace(&mut w.poll_state, PollState::SyncPolling);
+      assert!(prev_poll_state != PollState::SyncPolling);
```

Comment:
I wouldn't be surprised if this assertion would fail in certain edge cases, mostly because `handle_interrupt` will just directly call `poll_sessions_sync()`.

Something like:
* connect inspector
* hit pause
* run `for (;;) {}` in the devtools console
* hit stop

Reply:
> I wouldn't be surprised if this assertion would fail in certain edge cases, mostly because handle_interrupt will just directly call poll_sessions_sync().

But that would mean that `poll_sessions_sync` is reentrant - I don't think that's really possible, because this function will block.

I'll test this scenario though

Reply:
```js
// foo.js
console.log("hello");
```

```
$ cargo run -- run --inspect-brk foo.js
Debugger listening on ws://127.0.0.1:9229/ws/35c8fc10-adec-41e5-b865-fcc88c48e363
Visit chrome://inspect to connect to the debugger.
Deno is waiting for debugger to connect.
Debugger session started.

// Run for (;;) {}
// Resume execution
// Everything works fine
// Close Devtools

Debugger session ended

// Hit Ctrl + C to exit since infinite loop is running
```

Seems to be fine?

Reply:
> But that would mean that poll_sessions_sync is reentrant - I don't think that's really possible, because this function will block.

There's a chance that it won't happen in practice because the `Waker::inspector_ptr` field contains `None`, so no interrupt gets requested.

But I'm not 100% positive that V8 *won't* call `run_message_loop_on_pause()` in response to messages that are dispatched from `poll_sessions_sync()`. But let's try it out.

It's not a showstopper if this isn't perfect yet. IIRC in a previous iteration the inspector would just hang indefinitely if you ever tried this.

> I'll test this scenario though

👍